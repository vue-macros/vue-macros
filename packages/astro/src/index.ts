import { resolveOptions, type Options } from 'vue-macros'
import VueMacros from 'vue-macros/vite'
import type { AstroIntegration, ViteUserConfig } from 'astro'
import type { Plugin } from 'vite'

function findPluginAndRemove(
  name: string,
  plugins: ViteUserConfig['plugins'],
): Plugin | undefined {
  const idx = plugins!.findIndex(
    (plugin) => plugin && 'name' in plugin && plugin.name === name,
  )
  if (idx === -1) return
  const plugin = plugins![idx]
  plugins!.splice(idx, 1)
  return plugin as any
}

export default function (options?: Options): AstroIntegration {
  return {
    name: '@vue-macros/astro',
    hooks: {
      'astro:config:setup': async ({ config }) => {
        const resolvedOptions = await resolveOptions(options || {})
        const vue = findPluginAndRemove('vite:vue', config.vite.plugins)
        const vueJsx = findPluginAndRemove('vite:vue-jsx', config.vite.plugins)

        const vueMacrosPlugins = await VueMacros({
          ...resolvedOptions,
          plugins: {
            vue,
            vueJsx,
          },
        })
        config.vite.plugins ||= []
        config.vite.plugins.push(...vueMacrosPlugins)
      },
    },
  }
}
