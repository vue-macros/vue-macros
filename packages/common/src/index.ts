export * from 'magic-string-ast'
export * from 'ast-kit'
export {
  createFilter as createRollupFilter,
  normalizePath,
  type FilterPattern,
} from 'unplugin-utils'

export * from './ast'
export * from './constants'
export * from './dep'
export * from './options'
export * from './types'
export * from './unplugin'
export * from './vue'
