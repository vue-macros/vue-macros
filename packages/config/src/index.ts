export * from './config'
export * from './define'
export * from './options'
