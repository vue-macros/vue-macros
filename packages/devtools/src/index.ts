import { resolve } from 'node:path'
import getPort from 'get-port'
import sirv from 'sirv'
import type { Plugin } from 'vite'

const DEV_SERVER_PATH = '/__vue-macros'

export interface Options {
  /** @internal */
  nuxtContext?: {
    isClient?: boolean
  }
}

export const Devtools = ({ nuxtContext }: Options = {}): Plugin => {
  return {
    name: 'vue-macros-devtools',
    async configureServer(server) {
      if (nuxtContext?.isClient === false) return

      if (import.meta.DEV) {
        const { createServer } = await import('vite')
        const subServer = await createServer({
          root: resolve(import.meta.dirname, '../src/client'),
          server: {
            hmr: {
              port: await getPort(),
            },
            middlewareMode: true,
          },
        })
        server.middlewares.use(DEV_SERVER_PATH, subServer.middlewares)
      } else {
        server.middlewares.use(
          DEV_SERVER_PATH,
          sirv(resolve(import.meta.dirname, 'client'), {
            single: true,
            dev: true,
          }),
        )
      }
    },
  }
}
