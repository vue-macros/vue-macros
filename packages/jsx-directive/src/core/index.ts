import {
  babelParse,
  generateTransform,
  getLang,
  MagicStringAST,
  parseSFC,
  REGEX_SETUP_SFC,
  walkAST,
  type CodeTransform,
} from '@vue-macros/common'
import type { OptionsResolved } from '..'
import * as helper from './helper'
import { transformVFor } from './v-for'
import { transformVHtml } from './v-html'
import { transformVIf } from './v-if'
import { transformVMemo } from './v-memo'
import { transformVModel } from './v-model'
import { transformOnWithModifiers, transformVOn } from './v-on'
import { transformVSlot, type VSlotMap } from './v-slot'
import type { JSXAttribute, JSXElement, Node, Program } from '@babel/types'

export * from './restructure'
export const withDefaultsHelperCode: string = helper.withDefaultsHelperCode

export type JsxDirective = {
  node: JSXElement
  attribute: JSXAttribute
  parent?: Node | null
  vIfAttribute?: JSXAttribute
  vForAttribute?: JSXAttribute
  vMemoAttribute?: JSXAttribute
}

const onWithModifiersRegex = /^on[A-Z]\S*_\S+/

export function transformJsxDirective(
  code: string,
  id: string,
  options: OptionsResolved,
): CodeTransform | undefined {
  const lang = getLang(id)

  const programs: [program: Program, offset: number][] = []
  if (lang === 'vue' || REGEX_SETUP_SFC.test(id)) {
    const { scriptSetup, getSetupAst, script, getScriptAst } = parseSFC(
      code,
      id,
    )
    if (script) {
      programs.push([getScriptAst()!, script.loc.start.offset])
    }
    if (scriptSetup) {
      programs.push([getSetupAst()!, scriptSetup.loc.start.offset])
    }
  } else if (['jsx', 'tsx'].includes(lang)) {
    programs.push([babelParse(code, lang), 0])
  } else {
    return
  }

  const s = new MagicStringAST(code)
  for (const [ast, offset] of programs) {
    s.offset = offset
    transform(s, ast, options)
  }
  return generateTransform(s, id)
}

function transform(
  s: MagicStringAST,
  program: Program,
  options: OptionsResolved,
) {
  const { prefix, version } = options
  const vIfMap = new Map<Node | null | undefined, JsxDirective[]>()
  const vForNodes: JsxDirective[] = []
  const vMemoNodes: JsxDirective[] = []
  const vHtmlNodes: JsxDirective[] = []
  const vSlotMap: VSlotMap = new Map()
  const vOnNodes: JsxDirective[] = []
  const onWithModifiers: JsxDirective[] = []
  walkAST<Node>(program, {
    enter(node, parent) {
      if (node.type !== 'JSXElement') return
      const tagName = s.sliceNode(node.openingElement.name)

      let vIfAttribute
      let vForAttribute
      let vMemoAttribute
      let vSlotAttribute
      for (const attribute of node.openingElement.attributes) {
        if (attribute.type !== 'JSXAttribute') continue

        if (
          [`${prefix}if`, `${prefix}else-if`, `${prefix}else`].includes(
            String(attribute.name.name),
          )
        ) {
          vIfAttribute = attribute
        } else if (attribute.name.name === `${prefix}for`) {
          vForAttribute = attribute
        } else if (
          [`${prefix}memo`, `${prefix}once`].includes(
            String(attribute.name.name),
          )
        ) {
          vMemoAttribute = attribute
        } else if (attribute.name.name === `${prefix}html`) {
          vHtmlNodes.push({
            node,
            attribute,
          })
        } else if (
          (attribute.name.type === 'JSXNamespacedName'
            ? attribute.name.namespace
            : attribute.name
          ).name === `${prefix}slot`
        ) {
          vSlotAttribute = attribute
        } else if (attribute.name.name === `${prefix}on`) {
          vOnNodes.push({
            node,
            attribute,
          })
        } else if (onWithModifiersRegex.test(String(attribute.name.name))) {
          onWithModifiers.push({
            node,
            attribute,
          })
        } else if (
          attribute.name.type === 'JSXNamespacedName' &&
          attribute.name.namespace.name === `${prefix}model`
        ) {
          transformVModel(attribute, s, options)
        }
      }

      if (!vSlotAttribute || tagName !== 'template') {
        if (vIfAttribute) {
          vIfMap.get(parent) || vIfMap.set(parent, [])
          vIfMap.get(parent)!.push({
            node,
            attribute: vIfAttribute,
            parent,
          })
        }

        if (vForAttribute) {
          vForNodes.unshift({
            node,
            attribute: vForAttribute,
            vIfAttribute,
            parent,
            vMemoAttribute,
          })
        }
      }

      if (vMemoAttribute) {
        vMemoNodes.push({
          node,
          attribute: vMemoAttribute,
          parent: vForAttribute || vIfAttribute ? undefined : parent,
          vForAttribute,
        })
      }

      if (vSlotAttribute) {
        const slotNode = tagName === 'template' ? parent : node
        if (slotNode?.type !== 'JSXElement') return

        const attributeMap =
          vSlotMap.get(slotNode)?.attributeMap ||
          vSlotMap
            .set(slotNode, {
              vSlotAttribute:
                tagName !== 'template' ? vSlotAttribute : undefined,
              attributeMap: new Map(),
            })
            .get(slotNode)!.attributeMap
        const children =
          attributeMap.get(vSlotAttribute)?.children ||
          attributeMap
            .set(vSlotAttribute, {
              children: [],
              ...(tagName === 'template'
                ? {
                    vIfAttribute,
                    vForAttribute,
                  }
                : {}),
            })
            .get(vSlotAttribute)!.children

        if (slotNode === parent) {
          children.push(node)

          if (attributeMap.get(null)) return
          for (const child of parent.children) {
            if (
              (child.type === 'JSXElement' &&
                s.sliceNode(child.openingElement.name) === 'template') ||
              (child.type === 'JSXText' && !s.sliceNode(child).trim())
            )
              continue
            const defaultNodes =
              attributeMap.get(null)?.children ||
              attributeMap.set(null, { children: [] }).get(null)!.children
            defaultNodes.push(child)
          }
        } else {
          children.push(...node.children)
        }
      }
    },
  })

  vIfMap.forEach((nodes) => transformVIf(nodes, s, options))
  transformVFor(vForNodes, s, options)
  if (!version || version >= 3.2) transformVMemo(vMemoNodes, s, options)
  transformVHtml(vHtmlNodes, s)
  transformVOn(vOnNodes, s)
  transformOnWithModifiers(onWithModifiers, s, options)
  transformVSlot(vSlotMap, s, options)
}
