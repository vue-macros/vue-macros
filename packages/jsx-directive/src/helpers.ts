export { renderList, withKeys, withMemo, withModifiers } from 'vue'
