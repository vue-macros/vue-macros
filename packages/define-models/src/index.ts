import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  normalizePath,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { transformDefineModels } from './core'
import {
  emitHelperCode,
  emitHelperId,
  helperPrefix,
  useVmodelHelperCode,
  useVmodelHelperId,
} from './core/helper'

export interface Options extends BaseOptions {}

export type OptionsResolved = MarkRequired<Options, 'include' | 'version'>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      resolveId(id) {
        if (normalizePath(id).startsWith(helperPrefix)) return id
      },

      loadInclude(id) {
        return normalizePath(id).startsWith(helperPrefix)
      },

      load(_id) {
        const id = normalizePath(_id)
        if (id === emitHelperId) return emitHelperCode
        else if (id === useVmodelHelperId) return useVmodelHelperCode
      },

      transformInclude: filter,
      transform(code, id) {
        return transformDefineModels(code, id)
      },
    }
  },
)
export default plugin
