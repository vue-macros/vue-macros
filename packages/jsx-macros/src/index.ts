import { createUnplugin, type UnpluginInstance } from 'unplugin'
import { plugin, type Options, type OptionsResolved } from './core/plugin'

export type { Options, OptionsResolved }

const unplugin: UnpluginInstance<Options | undefined, false> =
  createUnplugin(plugin)
export default unplugin
