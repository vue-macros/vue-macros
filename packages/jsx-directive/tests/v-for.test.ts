import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformJsxDirective } from '../src/api'

describe('v-for', () => {
  describe('vue', async () => {
    await testFixtures(
      import.meta.glob<string>('./fixtures/v-for/*.{vue,jsx,tsx}', {
        eager: true,
        query: '?raw',
        import: 'default',
      }),
      (_, id, code) =>
        transformJsxDirective(code, id, {
          version: 3,
          lib: 'vue',
          prefix: 'v-',
        })?.code,
    )
  })

  describe('vue/vapor', async () => {
    await testFixtures(
      import.meta.glob<string>('./fixtures/v-for/*.{vue,jsx,tsx}', {
        eager: true,
        query: '?raw',
        import: 'default',
      }),
      (_, id, code) =>
        transformJsxDirective(code, id, {
          version: 3,
          lib: 'vue/vapor',
          prefix: 'v-',
        })?.code,
    )
  })
})
