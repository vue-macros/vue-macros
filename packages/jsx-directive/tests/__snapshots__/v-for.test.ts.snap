// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`v-for > vue > ./fixtures/v-for/index.vue 1`] = `
"<script setup lang="tsx">
import { renderList as __MACROS_renderList } from "vue";
const map = new Map([
  [1, '2'],
  [3, '4'],
])
const set = new Set(['1', '2', '3'])
const object = { id: 1, name: 'admin' }
let selected = 0

const Comp = <>{__MACROS_renderList(4, (i) => <div>{i}</div>)}</>

defineRender(() => (
  <>
    {__MACROS_renderList(4, (i) => <div key={i}>
      <div>{i}</div>
    </div>)}

    {__MACROS_renderList(object, (value, key, index) => <div key={index}>
      {key}: {value}
    </div>)}

    {__MACROS_renderList([1, 2, 3][Symbol.iterator](), (i, index) => <div key={index}>
      <div>{i}</div>
    </div>)}

    {__MACROS_renderList([1, 2, 3], (i, __MACROS_index) => <div v-memo={[selected === i]} key={i}>
      <div>{i}</div>
    </div>)}

    {(set) ? __MACROS_renderList(set, (i) => <div key={i}>
      <div>{i}</div>
    </div>) : null}

    {__MACROS_renderList(map, ([key, value], index) => <div key={index}>
      {key}
      {__MACROS_renderList(3, (key) => <div>
        {key}: {value}
      </div>)}
    </div>)}
  </>
))
</script>
"
`;

exports[`v-for > vue/vapor > ./fixtures/v-for/index.vue 1`] = `
"<script setup lang="tsx">
import { createFor as __MACROS_createFor } from "vue";
const map = new Map([
  [1, '2'],
  [3, '4'],
])
const set = new Set(['1', '2', '3'])
const object = { id: 1, name: 'admin' }
let selected = 0

const Comp = <>{__MACROS_createFor(() => 4, (...__MACROS_props) => <div>{__MACROS_props[0].value}</div>)}</>

defineRender(() => (
  <>
    {__MACROS_createFor(() => 4, (...__MACROS_props) => <div>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i]) => (i))}

    {__MACROS_createFor(() => object, (...__MACROS_props) => <div>
      {__MACROS_props[1].value}: {__MACROS_props[0].value}
    </div>, ([value, key, index]) => (index))}

    {__MACROS_createFor(() => [1, 2, 3][Symbol.iterator](), (...__MACROS_props) => <div>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i, index]) => (index))}

    {__MACROS_createFor(() => [1, 2, 3], (...__MACROS_props) => <div v-memo={[selected === __MACROS_props[0].value]}>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i, __MACROS_index]) => (i))}

    {(set) ? __MACROS_createFor(() => set, (...__MACROS_props) => <div>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i]) => (i)) : null}

    {__MACROS_createFor(() => map, (...__MACROS_props) => <div>
      {__MACROS_props[0].value[0].value}
      {__MACROS_createFor(() => 3, (...__MACROS_props) => <div>
        {__MACROS_props[0].value}: {__MACROS_props[0].value[1].value}
      </div>)}
    </div>, ([[key, value], index]) => (index))}
  </>
))
</script>
"
`;
